src/compiler/scala/tools/nsc/profile/Profiler.scala

```diff
@@ -0,0 +1,153 @@
+        new StreamProfileReporter(new PrintWriter(new FileWriter(settings.YprofileDestination.value, true)))
+      else ConsoleProfileReporter
+      new RealProfiler(reporter, settings)
+    }
+}
+
+
+case class ProfileCounters(wallClockTimeNanos: Long, cpuTimeNanos: Long, userTimeNanos: Long, allocatedBytes: Long,
+                           retainedHeapBytes: Long, gcTimeMillis: Long) {
+  def -(that: ProfileCounters): ProfileCounters = {
+    ProfileCounters(
+      this.wallClockTimeNanos - that.wallClockTimeNanos,
+      this.cpuTimeNanos - that.cpuTimeNanos,
+      this.userTimeNanos - that.userTimeNanos,
+      this.allocatedBytes - that.allocatedBytes,
+      this.retainedHeapBytes - that.retainedHeapBytes,
+      this.gcTimeMillis - that.gcTimeMillis)
+  }
+  def updateHeap(heapDetails: ProfileCounters): ProfileCounters = {
+    copy(retainedHeapBytes = heapDetails.retainedHeapBytes)
+  }
+  private def toMillis(ns: Long) = ns / 1000000.0D
+  private def toMegaBytes(bytes: Long) = bytes / 1000000.0D
+
+  def wallClockTimeMillis: Double = toMillis(wallClockTimeNanos)
+  def cpuTimeMillis: Double = toMillis(cpuTimeNanos)
+  def userTimeMillis: Double = toMillis(userTimeNanos)
+  def allocatedMB: Double = toMegaBytes(allocatedBytes)
+  def retainedHeapMB: Double = toMegaBytes(retainedHeapBytes)
+
+}
+
+sealed trait Profiler {
+  def finished(): Unit
+
+  def after(phase: Phase, profileBefore: ProfileCounters): Unit
+
+  def before(phase: Phase): ProfileCounters
+
+}
+
+private[profile] object NoOpProfiler extends Profiler {
+  val noSnap = ProfileCounters(0, 0, 0, 0, 0, 0)
+
+  override def before(phase: Phase): ProfileCounters = noSnap
+
+  override def after(phase: Phase, profileBefore: ProfileCounters): Unit = ()
+
+  override def finished(): Unit = ()
+}
+
+private[profile] object RealProfiler {
+  import scala.collection.JavaConverters._
+  private val runtimeMx = ManagementFactory.getRuntimeMXBean
+  private val memoryMx = ManagementFactory.getMemoryMXBean
+  private val gcMx = ManagementFactory.getGarbageCollectorMXBeans.asScala.toList
+  private val threadMx = ExtendedThreadMxBean.proxy
+  if (threadMx.isThreadCpuTimeSupported) threadMx.setThreadCpuTimeEnabled(true)
+  private val idGen = new AtomicInteger()
+}
+
+private[profile] class RealProfiler(reporter: ProfileReporter, val settings: Settings) extends Profiler {
+  def outDir: String = settings.outdir.value
+  val id: Int = RealProfiler.idGen.incrementAndGet()
+
+  private def snap: ProfileCounters = {
+    import RealProfiler._
+    ProfileCounters(System.nanoTime(), threadMx.getCurrentThreadCpuTime, threadMx.getCurrentThreadUserTime,
+      threadMx.getThreadAllocatedBytes(Thread.currentThread().getId), memoryMx.getHeapMemoryUsage.getUsed,
+      gcMx.foldLeft(0L) { case (sum, bean) => bean.getCollectionTime + sum })
+  }
+  private val overhead = {
+    val s1 = snap
+    val s2 = snap
+    s2 - s1
+  }
+  private def doGC(): Unit = {
+    System.gc()
```

Comment:
for the record - it also has inaccuracies related to finalization and object graphs that require multiple GC/finalization cycles to reclaim, and also with soft references